EmailValidator/EmailLexer.php

```diff
@@ -77,10 +77,16 @@ class EmailLexer extends AbstractLexer
 
     protected $previous;
 
+    public function __construct()
+    {
+        $this->reset();
```

Comment:
Calling `reset` on construction sounds a bit odd, reseting something 'new'. For what I've seen in the rest of the changes, this is needed mainly to do the equality evaluation (e.g https://github.com/egulias/EmailValidator/pull/203/files#diff-c759db088d5c729c1f537d85c2ea93d0R187).
I would suggest something like `init()` :
```php
$this->hasInvalidTokens = false;
$this->previous = $this->token = ['type' => null, 'value' => '', 'position' => 0];
```
and call that from both `__contruct` & `reset`.

Reply:
the code is full of checks with `$this->lexer->token['type']`, when `$this->lexer->token` is `null` actually. PHP 7.4 is now refusing to access keys on nul values, so that the affected properties must be set to an array representing the null-state. reset does it.

Reply:
Thanks for the explanation, understood. My suggestion is moving both lines to a new method so they can call it having a more expresive meaning and avoiding coupling too construction with re-use (reseting to analyse another string).

```php
__construct()
{
    $this->init();
}

reset()
{
...
$this->init();
}
```